base_projects/Drivetrain/src/main/cpp/Robot.cpp

```diff
@@ -0,0 +1,367 @@
+        } else {
+          m_gyro =
+              std::make_unique<frc::ADXRS450_Gyro>(frc::SPI::Port::kOnboardCS0);
+        }
+        // } else if (gyroType == "NavX") {
+        //   if (gyroCtor == "SerialPort.kUSB") {
+        //     m_gyro = std::make_unique<AHRS>(frc::SerialPort::Port::kUSB);
+        //   } else if (gyroCtor == "I2C") {
+        //     m_gyro = std::make_unique<AHRS>(frc::I2C::Port::kMXP);
+        //   } else if (gyroCtor == "SerialPort.kMXP") {
+        //     m_gyro = std::make_unique<AHRS>(frc::SerialPort::Port::kMXP);
+        //   } else {
+        //     m_gyro = std::make_unique<AHRS>(frc::SPI::Port::kMXP);
+        //   }
+      } else if (gyroType == "Romi") {
+        m_gyro = std::make_unique<frc::RomiGyro>();
+      } else {
+        std::stringstream ss(gyroCtor);
+        int port = 0;
+        ss >> port;
+        m_gyro = std::make_unique<frc::AnalogGyro>(port);
+      }
+
+      m_gyroPosition = [&, this] {
+        return units::radian_t(units::degree_t{m_gyro->GetAngle()})
+            .to<double>();
+      };
+
+      m_gyroRate = [&, this] {
+        return units::radians_per_second_t(
+                   units::degrees_per_second_t{m_gyro->GetAngle()})
+            .to<double>();
+      };
+    }
+  } catch (std::exception& e) {
+    wpi::outs() << "Project failed: " << e.what() << "\n";
+    std::exit(-1);
+  }
+
+#ifdef INTEGRATION
+  // TODO use std::exit or EndCompetition once CTRE bug is fixed
+  std::set_terminate([]() { std::_Exit(0); });
+#endif
+}
+
+void Robot::AddMotorController(
+    int port, std::string controller, bool inverted,
+    std::vector<std::unique_ptr<frc::SpeedController>>* controllers) {
+  if (controller == "TalonSRX" || controller == "VictorSPX" ||
+      controller == "TalonFX") {
+    if (controller == "TalonSRX") {
+      controllers->emplace_back(std::make_unique<WPI_TalonSRX>(port));
+    } else if (controller == "TalonFX") {
+      controllers->emplace_back(std::make_unique<WPI_TalonFX>(port));
+    } else {
+      controllers->emplace_back(std::make_unique<WPI_VictorSPX>(port));
+    }
+    dynamic_cast<WPI_BaseMotorController*>(controllers->back().get())
+        ->SetInverted(inverted);
+  } else if (controller == "SPARK MAX (Brushless)" ||
+             controller == "SPARK MAX (Brushed)") {
+    if (controller == "SPARK MAX (Brushless)") {
+      controllers->emplace_back(std::make_unique<rev::CANSparkMax>(
+          port, rev::CANSparkMax::MotorType::kBrushless));
+    } else {
+      controllers->emplace_back(std::make_unique<rev::CANSparkMax>(
+          port, rev::CANSparkMax::MotorType::kBrushed));
+    }
+    static_cast<rev::CANSparkMax*>(controllers->back().get())
+        ->SetInverted(inverted);
+  } else {
+    controllers->emplace_back(std::make_unique<frc::Spark>(port));
+    static_cast<frc::Spark*>(controllers->back().get())->SetInverted(inverted);
+  }
+}
+
+void Robot::SetupEncoders(const std::string& controllerName,
```

Comment:
I think this is the right thing? It's either `const wpi::Twine&` or `wpi::StringRef.
```suggestion
void Robot::SetupEncoders(const wpi::Twine& controllerName,
```